### PR TITLE
feat: change data model loads rated power values cos phi to array

### LINKED ISSUE #80: feat: change data model loads rated power values cos phi to array

### PR DESCRIPTION
**Pull request by sasanjac**
Fixes #80 #84

### CODE REVIEW

File: powerfactory_tools/schema/steadystate_case/active_power.py [hunk 1]
```diff
@@ -17,16 +17,24 @@
...
-            msg = f"Power mismatch: Total reactive power should be {p_total}, is {values['value_0']}."
+            msg = f"Power mismatch: Total active power should be {p_total}, is {values['value_0']}."
+            raise ValueError(msg)
+
+        return values
+
+    @pydantic.root_validator
+    def validate_symmetry(cls, values: dict[str, Any]) -> dict[str, Any]:
+        if values["is_symmetrical"] and not (values["value_a_0"] == values["value_b_0"] == values["value_c_0"]):
+            msg = "Power mismatch: Active power is not symmetrical."
```
Comment (SebastianDD) [inline powerfactory_tools/schema/steadystate_case/active_power.py:37]:
```suggestion
            msg = "Power mismatch: Active power of load is symmetrical, but active power of single phases are not symmetrical."
```

File: powerfactory_tools/schema/steadystate_case/reactive_power.py [hunk 2]
```diff
@@ -18,21 +18,25 @@
...
-            msg = f"Power mismatch: Total reactive values should be {q_total}, is {values['value_0']}."
+            msg = f"Power mismatch: Total reactive power should be {q_total}, is {values['value_0']}."
+            raise ValueError(msg)
+
+        return values
+
+    @pydantic.root_validator
+    def validate_symmetry(cls, values: dict[str, Any]) -> dict[str, Any]:
+        if values["is_symmetrical"] and not (values["value_a_0"] == values["value_b_0"] == values["value_c_0"]):
+            msg = "Power mismatch: Rective power is not symmetrical."
```
Comment (SebastianDD) [inline powerfactory_tools/schema/steadystate_case/reactive_power.py:39]:
```suggestion
            msg = "Power mismatch: Rective power of load is symmetrical, but reactive power of single phases are not symmetrical."
```

File: powerfactory_tools/schema/topology/load.py [hunk 3]
```diff
@@ -70,15 +70,48 @@ class ConsumerSystemType(Enum):
...
-    value_s: float  # rated apparent power (phase s)
-    value_t: float  # rated apparent power (phase t)
-    cosphi: float | None  # rated cos(phi) in relation to rated power
-    cosphi_r: float | None  # rated cos(phi) (phase r)
-    cosphi_s: float | None  # rated cos(phi) (phase s)
-    cosphi_t: float | None  # rated cos(phi) (phase t)
+    value_a: float  # rated apparent power (phase a)
+    value_b: float  # rated apparent power (phase b)
+    value_c: float  # rated apparent power (phase c)
+    cosphi_ue: float = 1  # rated cos(phi) in relation to rated power
```
Comment (SebastianDD) [inline powerfactory_tools/schema/topology/load.py:94]:
```suggestion
    cosphi_ue: float = 1  # rated cos(phi) in relation to rated power [under excited]
```

File: powerfactory_tools/schema/topology/load.py [hunk 4]
```diff
@@ -70,15 +70,48 @@ class ConsumerSystemType(Enum):
...
+    cosphi_a_ue: float = 1  # rated cos(phi) (phase a)
+    cosphi_b_ue: float = 1  # rated cos(phi) (phase b)
+    cosphi_c_ue: float = 1  # rated cos(phi) (phase c)
+    cosphi_oe: float = 1  # rated cos(phi) in relation to rated power
```
Comment (SebastianDD) [inline powerfactory_tools/schema/topology/load.py:98]:
```suggestion
    cosphi_oe: float = 1  # rated cos(phi) in relation to rated power [over excited]
```

File: powerfactory_tools/schema/steadystate_case/reactive_power.py (hunk 2 above)
Comment (sasanjac) [inline powerfactory_tools/schema/steadystate_case/reactive_power.py:39]:
```suggestion
            msg = "Power mismatch: Three-phase reactive power of load is not symmetrical."
```

File: powerfactory_tools/schema/steadystate_case/active_power.py (hunk 1 above)
Comment (sasanjac) [inline powerfactory_tools/schema/steadystate_case/active_power.py:37]:
```suggestion
            msg = "Power mismatch: Three-phase active power of load is not symmetrical."
```

File: tests/test_topology_load.py [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+            "cosphi_oe",
+            "cosphi_a_oe",
+            "cosphi_b_oe",
+            "cosphi_c_oe",
+            "expectation",
+        ),
+        [
+            (0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, does_not_raise()),
+            (1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, does_not_raise()),
+            (2, 2, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, does_not_raise()),
```
Comment (SebastianDD) [inline tests/test_topology_load.py:33]: shouldn't raise Validation Error?

Comment (sasanjac) [reply tests/test_topology_load.py]: No, those are the power values

File: tests/test_topology_load.py [hunk 6]
```diff
@@ -17,22 +17,18 @@ class TestRatedPower:
...
         ),
         [
-            (0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, does_not_raise()),
-            (1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, does_not_raise()),
-            (2, 2, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, does_not_raise()),
-            (2, 2, 1, 1, 1, 1, 2, 1, 1, 1, 1, 1, pytest.raises(pydantic.ValidationError)),
-            (2, -2, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, pytest.raises(pydantic.ValidationError)),
+            (0, 0, 0, 0, 0, 0, 0, 0, does_not_raise()),
+            (1, 1, 1, 1, 1, 1, 1, 1, does_not_raise()),
+            (2, 2, 1, 1, 1, 1, 1, 1, does_not_raise()),
```
Comment (SebastianDD) [inline tests/test_topology_load.py:29]: so this test does not check if s=s_a+s_b+s_c ?

Comment (sasanjac) [reply tests/test_topology_load.py:29]: `RatedPower` does not have a symmetry check

Comment (SebastianDD) [reply tests/test_topology_load.py:29]: @sasanjac : should we add validator for class rated power to check if symmetrical (like for active_power_ssc --> cfg. test/active_power)

**Comment by SebastianDD** [comment]
#85 is still unfixed.

example for asymmetrical loads:
```
"name": "LV_Load_7a_FIXED",
      "reactive_power": {
        "controller": null,
        "is_symmetrical": false,
        "value_0": 3098.0,
        "value_a_0": 1006.64,
        "value_b_0": 1153.78,
        "value_c_0": 937.44
      }
```
`q_total = 3097.86`; so `diff = 3098-9097.86 = 0.14 > Threshold (0.01)`

* asym: The maximum rounding error calculating value_0 from value_a,b,c is 0.5
* sym: The maximum rounding error calculating value_a,b,c from value_0 is 0.01

--> Threshold should be 0.5?

File: tests/test_topology_load.py (hunk 6 above)
Comment (sasanjac) [reply tests/test_topology_load.py:29]: I actually don't think that we have this information. We either have three single phase rated powers or one three phase rated power.

**Review by SebastianDD — APPROVED** [review]

**Pull request merged**